### ISSUE TITLE
GRPO colocate模式有办法不offload奖励模型吗

### ISSUE THREAD

**Issue by warriorandroyal** [modelscope/ms-swift#5585]
八卡H100训练，脚本如下，我发现训练的时候会很慢，每一步都需要重新加载奖励模型特别慢（一次大概五分钟），奖励模型是3B的qwen2.5训练的，有没有可能只卸载主模型不卸载奖励模型
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
swift rlhf \
    --rlhf_type grpo \
    --model main_model_sft_v3
    --model_type qwen2_5 \
    --external_plugins examples/train/grpo/plugin/plugin.py \
    --reward_funcs external_mutisummary_format external_mutisummary_accuracy soft_overlong\
    --reward_model reward_model_sft_v3\
    --reward_model_plugin summary_rm \
    --overlong_filter True \
    --dynamic_sample True \
    --soft_cache_length 1024 \
    --epsilon_high 0.28 \
    --reward_weights 1 1 1 1\
    --loss_type bnpo \
    --max_resample_times 3 \
    --add_version False \
    --run_name main_dapo\
    --project_name summary\
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.5 \
    --vllm_tensor_parallel_size 4 \
    --offload_optimizer true \
    --offload_model true \
    --vllm_max_model_len 12000 \
    --sleep_level 1 \
    --train_type full \
    --torch_dtype bfloat16 \
    --dataset main_model_dapo_data.jsonl \
    --logging_dir "$logging_dir" \
    --max_length 12000 \
    --max_completion_length 3072 \
    --truncation_strategy left \
    --num_train_epochs 3 \
    --max_epochs 3 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 2 \
    --eval_steps 200 \
    --save_steps 200 \
    --save_total_limit 5 \
    --logging_steps 1 \
    --output_dir "$output_dir" \
    --warmup_ratio 0.01 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --split_dataset_ratio 0.01\
    --deepspeed zero3 \
    --log_completions true \
    --log_entropy true \
    --report_to wandb \
    --beta 0.04\
    --temperature 1.1 \
    --top_p 1.0 \
    --top_k 80 \

**Comment by hjh0119** [comment]
> 每一步都需要重新加载奖励模型特别慢

offload奖励模型是从哪看出来的

**Comment by warriorandroyal** [comment]
0%|          | 0/4 [00:00<?, ?it/s][A
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]

100%|██████████| 4/4 [05:28<00:00, 82.05s/it]


100%|██████████| 4/4 [05:25<00:00, 81.39s/it]

100%|██████████| 4/4 [05:28<00:00, 82.02s/it]

100%|██████████| 4/4 [05:28<00:00, 82.04s/it]
100%|██████████| 4/4 [05:28<00:00, 82.04s/it]
100%|██████████| 4/4 [05:25<00:00, 81.38s/it]

100%|██████████| 4/4 [05:25<00:00, 81.38s/it]

100%|██████████| 4/4 [05:28<00:00, 82.06s/it][A
100%|██████████| 4/4 [05:28<00:00, 82.06s/it]
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.

100%|██████████| 4/4 [05:25<00:00, 81.37s/it]
100%|██████████| 4/4 [05:25<00:00, 81.37s/it]

Train:   0%|          | 1/6702 [06:09<688:36:56, 369.95s/it]
                                                            

Train:   0%|          | 1/6702 [06:09<688:36:56, 369.95s/it]
Train:   0%|          | 1/6702 [06:09<688:36:56, 369.95s/it]

  0%|          | 0/4 [00:00<?, ?it/s][A
  0%|          | 0/4 [00:00<?, ?it/s]

100%|██████████| 4/4 [06:00<00:00, 90.18s/it]

100%|██████████| 4/4 [06:02<00:00, 90.63s/it]

100%|██████████| 4/4 [06:02<00:00, 90.64s/it]
100%|██████████| 4/4 [06:02<00:00, 90.64s/it]
100%|██████████| 4/4 [06:02<00:00, 90.64s/it]

100%|██████████| 4/4 [06:00<00:00, 90.18s/it]
100%|██████████| 4/4 [06:02<00:00, 90.64s/it]


100%|██████████| 4/4 [06:00<00:00, 90.18s/it]

100%|██████████| 4/4 [06:00<00:00, 90.11s/it]
100%|██████████| 4/4 [06:02<00:00, 90.64s/it][A
100%|██████████| 4/4 [06:00<00:00, 90.11s/it]

100%|██████████| 4/4 [06:02<00:00, 90.64s/it]
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.

Train:   0%|          | 2/6702 [12:46<717:11:11, 385.35s/it]
                                                            

Train:   0%|          | 2/6702 [12:46<717:11:11, 385.35s/it]
Train:   0%|          | 2/6702 [12:46<717:11:11, 385.35s/it]

  0%|          | 0/4 [00:00<?, ?it/s][A
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]


100%|██████████| 4/4 [05:38<00:00, 84.75s/it]
100%|██████████| 4/4 [05:38<00:00, 84.75s/it]
100%|██████████| 4/4 [05:38<00:00, 84.75s/it]

100%|██████████| 4/4 [05:43<00:00, 85.90s/it]
100%|██████████| 4/4 [05:43<00:00, 85.89s/it]
100%|██████████| 4/4 [05:38<00:00, 84.75s/it]
100%|██████████| 4/4 [05:43<00:00, 85.90s/it]


100%|██████████| 4/4 [05:43<00:00, 85.89s/it]

100%|██████████| 4/4 [05:38<00:00, 84.68s/it]
100%|██████████| 4/4 [05:38<00:00, 84.74s/it]
100%|██████████| 4/4 [05:43<00:00, 85.90s/it]
100%|██████████| 4/4 [05:38<00:00, 84.68s/it]

100%|██████████| 4/4 [05:38<00:00, 84.74s/it]
[A
100%|██████████| 4/4 [05:43<00:00, 85.90s/it]
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.


  0%|          | 0/4 [00:00<?, ?it/s][A
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]

100%|██████████| 4/4 [05:31<00:00, 82.81s/it]
100%|██████████| 4/4 [05:29<00:00, 82.35s/it]
100%|██████████| 4/4 [05:31<00:00, 82.81s/it]

100%|██████████| 4/4 [05:29<00:00, 82.35s/it]

100%|██████████| 4/4 [05:29<00:00, 82.35s/it]

100%|██████████| 4/4 [05:29<00:00, 82.36s/it]
100%|██████████| 4/4 [05:29<00:00, 82.35s/it]
100%|██████████| 4/4 [05:29<00:00, 82.36s/it]


100%|██████████| 4/4 [05:29<00:00, 82.35s/it]
100%|██████████| 4/4 [05:31<00:00, 82.81s/it]
100%|██████████| 4/4 [05:31<00:00, 82.81s/it]

100%|██████████| 4/4 [05:29<00:00, 82.35s/it]

100%|██████████| 4/4 [05:31<00:00, 82.81s/it][A
100%|██████████| 4/4 [05:31<00:00, 82.81s/it]
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.
[WARNING:swift] Unable to extract reward score from the model's output (missing <answer> tag). Setting reward to 0.0.

Train:   0%|          | 3/6702 [24:53<1007:44:45, 541.56s/it]
                                                             

Train:   0%|          | 3/6702 [24:53<1007:44:45, 541.56s/it]
Train:   0%|          | 3/6702 [24:53<1007:44:45, 541.56s/it]

  0%|          | 0/4 [00:00<?, ?it/s][A
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]

100%|██████████| 4/4 [05:15<00:00, 78.97s/it]

100%|██████████| 4/4 [05:17<00:00, 79.35s/it]
100%|██████████| 4/4 [05:17<00:00, 79.35s/it]

100%|██████████| 4/4 [05:17<00:00, 79.28s/it]
100%|██████████| 4/4 [05:17<00:00, 79.34s/it]
100%|██████████| 4/4 [05:17<00:00, 79.28s/it]

100%|██████████| 4/4 [05:17<00:00, 79.34s/it]
100%|██████████| 4/4 [05:15<00:00, 78.98s/it]

100%|██████████| 4/4 [05:15<00:00, 78.97s/it]

100%|██████████| 4/4 [05:15<00:00, 78.97s/it]

100%|██████████| 4/4 [05:15<00:00, 78.98s/it]

100%|██████████| 4/4 [05:17<00:00, 79.32s/it][A
100%|██████████| 4/4 [05:17<00:00, 79.32s/it]

因为我发现每一步都要重新加载？

**Comment by hjh0119** [comment]
没看出来offload reward model的信息 offload_model 只控制训练模型和参考模型 

另外PT engine推理genrm比较慢 建议vllm部署 
https://swift.readthedocs.io/zh-cn/latest/Instruction/GRPO/DeveloperGuide/%E5%A5%96%E5%8A%B1%E6%A8%A1%E5%9E%8B.html#id4

**Comment by warriorandroyal** [comment]
我的意思是一直出现重新加载奖励模型的进度条，原因不应该是每次都会offload才要重新加载吗

**Comment by hjh0119** [comment]
这个进度条不是推理的进度条吗:(

**Comment by warriorandroyal** [comment]
可能是我理解错了，我vllm部署试试，感谢🌹
